etl/topline_summary.kp/knowledge.md

```diff
@@ -0,0 +1,447 @@
+
+### Execution
+Finally, the entry point for execution can be found near the end of the notebook. This is where the report date and period can be manipulated to generate a new dataset for the executive report. 
+
+## Setup and Helper Functions
+
+
+```python
+!pip install arrow --quiet
+```
+
+```python
+import logging
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(__name__)
+
+# reduce noise
+logging.getLogger('boto3').setLevel(logging.CRITICAL)
+logging.getLogger('botocore').setLevel(logging.CRITICAL)
+```
+This script takes a report start date, and will generate a report monthly or weekly. We need to get the week or month following the start date. If we are determining the count of inactive users, we will need to query in the [previous period](https://github.com/mozilla-services/data-pipeline/blob/master/reports/executive_summary/run_executive_report.py#L163).
+
+
+```python
+import arrow
+
+def date_range(start_date, period, last_period=False):
+    """Returns the begin_date, end_date for this report.
+    
+    start_date:  datestring in the format YYYYMMDD
+    period:      type of period, either 'months' or 'weeks'
+    last_period: determines if this will this or last month, defaults to this month
+    """
+    if period not in ('months', 'weeks', 'day'):
+        log.warning("{} is an invalid argument to date_range, defaulting to `weeks`"
+                    .format(period))
+        period = 'weeks'
+    
+    # Not used in production, this is for testing against a smaller subset of days
+    if period == 'day':
+        return start_date, start_date
+
+    fmt = 'YYYYMMDD'
+    begin_date = arrow.get(start_date, fmt)
+    if last_period:
+        # offset the start date by a single period
+        offset = {period: -1}
+        begin_date = begin_date.replace(**offset)
+
+    offset = {period: 1}
+    end_date = begin_date.replace(**offset)
+    
+    return begin_date.format(fmt), end_date.format(fmt)
+```
+### User Defined Functions
+These UDFs capture the logic described in the [`extract_executive_summary.lua` heka decoder]( https://github.com/mozilla-services/data-pipeline/blob/515b85101e7335a66fb8c26316cd9721c832b098/heka/sandbox/decoders/extract_executive_summary.lua).
+
+
+```python
+import re
+from functools import partial
+from pyspark.sql.functions import udf
+from pyspark.sql.types import *
+
+
+SEC_IN_HOUR = 60 * 60
+SEC_IN_DAY = SEC_IN_HOUR * 24
+
+# UDF for aggregating search counts based on search engine
+
+search_schema = StructType([
+    StructField("google", IntegerType(), False),
+    StructField("bing", IntegerType(), False),
+    StructField("yahoo", IntegerType(), False),
+    StructField("other", IntegerType(), False)])
+
+search_pat = ['[Gg]oogle', '[Bb]ing', '[Yy]ahoo', '.']
+search_pat_comp = [re.compile(pat) for pat in search_pat]
+
+def get_search_counts(row):
```

Comment:
These helper functions would be good candidates to move over to `standards.py` in [python_moztelemetry](https://github.com/mozilla/python_moztelemetry).